### PR TITLE
Add sshserver file transfer handlers

### PR DESCRIPTION
**Pull request by avatus**
Part of https://github.com/gravitational/teleport/pull/23546

This PR will add the handlers to the sshserver to listen for File Transfer Request events. There is a bit of refactor as well from https://github.com/gravitational/teleport/pull/24260, but the main thing being that file transfers over SFTP does not include an `ssh.Request` so I removed the comparison check. 

partner PR (that also includes web) https://github.com/gravitational/teleport/pull/24581

### CODE REVIEW

File: lib/session/session.go
Comment (capnspacehook) [inline lib/session/session.go:129]:
```suggestion
	// Requester is the authenticated Teleport user who requested the file transfer
	Requester string `json:"requester"`
```

File: lib/srv/sess.go
Comment (capnspacehook) [inline lib/srv/sess.go:432]: Shouldn't this check be outside the `for` loop?

File: lib/srv/sess.go
Comment (capnspacehook) [inline lib/srv/sess.go:1527]: Nit: could this be a `bool` or typed constant instead as the number of valid values here is very limited (2)?

File: lib/srv/termhandlers.go
Comment (capnspacehook) [inline lib/srv/termhandlers.go:279]: Is this checked anywhere to ensure it's either `upload` or `download`?

File: lib/session/session.go
Comment (zmb3) [inline lib/session/session.go:113]: This is a bit misleading, because it looks like this can be a request or a response based on the fields below.

File: lib/session/session.go
Comment (zmb3) [inline lib/session/session.go:123]: Would `Path` be a better name?

File: lib/session/session.go
Comment (zmb3) [inline lib/session/session.go:127]: What would this look like? Why a string?

File: lib/session/session.go
Comment (zmb3) [inline lib/session/session.go:131]: Why do you need an `Approved` field if you have approvers? Is it possible for `Approved` to be false but `Approvers` to be non-empty?

File: lib/session/session.go
Comment (zmb3) [reply lib/session/session.go:136]: Yeah, this doesn't really look like an Unmarshal operation to me.

File: lib/srv/regular/sshserver.go
Comment (zmb3) [inline lib/srv/regular/sshserver.go:1605]: Is this handling a request, a response, or both? The name makes it hard to tell.

File: lib/srv/sess.go
Comment (zmb3) [inline lib/srv/sess.go:394]:
```suggestion
// The notification is a global ssh request and requires the client to update its UI state accordingly.
```

File: lib/srv/sess.go
Comment (zmb3) [inline lib/srv/sess.go:399]: Should this be an error or is it expected that there will sometimes not be a session?

File: lib/srv/sess.go
Comment (zmb3) [reply lib/srv/sess.go:432]: Yes, almost certainly. Good catch.

File: lib/srv/sess.go
Comment (zmb3) [inline lib/srv/sess.go:410]: Aren't the events in `apievents` for the audit log?

File: lib/srv/sess.go
Comment (zmb3) [inline lib/srv/sess.go:1635]: So what happens in this scenario?

- 2 moderator approvals are required for transfer
- 3 moderators are watching the session
- moderator A approves the transfer
- moderator B denies the transfer

Does the request immediately get dropped, or does moderator C still have a chance to approve?

File: lib/session/session.go
Comment (avatus) [reply lib/session/session.go:113]: I agree. A comment on the other PR lead me to separate these from requests and responses so it will change (not sure outcome yet, but will update comment regardless)

File: lib/srv/termhandlers.go
Comment (avatus) [reply lib/srv/termhandlers.go:279]: I switched to an enum in the file transfer event and didnt switch here by mistake. I'm going to convert this to an enum so we won't have to do the check

File: lib/session/session.go
Comment (avatus) [reply lib/session/session.go:123]: i was keeping it in line with the presentation on the frontend and current file transfer (not request) params. Path does seem like a better name but a change for all the other aspects that aren't in this PR might muddy it up. I'll double check, and if its too many small changed, I'll separate into another PR

File: lib/session/session.go
Comment (avatus) [reply lib/session/session.go:127]: This was vestigial from testing/tinkering. will remove as it's not used

File: lib/session/session.go
Comment (avatus) [reply lib/session/session.go:131]: This is a bad consequence of having the request/response in this single struct. What `Approved` actually meant was "is this response and approval or deny", which now that I've seen through the weeds is NOT very clear. Removing it from this and making a discrete `Response` struct will clear it up

File: lib/srv/sess.go
Comment (avatus) [reply lib/srv/sess.go:410]: I haven't added the audit event yet, but a similar pattern with `Resize` is to create the audit event and send it, and then reuse that to send to the client which I liked. It looks more weird since the audit event isn't happening yet.

File: lib/srv/sess.go
Comment (avatus) [reply lib/srv/sess.go:1635]: Request will immediately be dropped. Similar to access requests, if moderated A approved and B denied, C won't be able to answer as the result is already in a `denied` state.

File: lib/srv/sess.go
Comment (avatus) [reply lib/srv/sess.go:1527]: I'll change this to a bool but also change the name to `Download`, that way its `Download = true` rather than `Direction = true` (doesn't make much sense).

File: lib/session/session.go
Comment (avatus) [reply lib/session/session.go:136]: I copied the name from the `UnmarshalTerminalParams` which i guess doesn't really fit into here now that you mention it. I'll update the name

File: lib/srv/sess.go
Comment (avatus) [reply lib/srv/sess.go:399]: Should be an error, thanks!

File: lib/session/session.go
Comment (zmb3) [inline lib/session/session.go:134]: Don't forget to update the godoc to account for the rename (here and below)

Comment (avatus) [reply lib/session/session.go]: I don't even think these methods are needed anymore. If i'm not checking/setting any defaults and just making a new struct, I'll just do it at the callsite and remove these all together.

File: lib/sshutils/req.go
Comment (avatus) [reply lib/sshutils/req.go:170]: 👍 to adding the domain. Thanks for the links

I see it being a bit confusing but I'm having trouble thinking of an answer that _isn't_ confusing which i think is inherent to this message being a "response". I could change it to maybe `file-transfer-request-approval` but that might not indicate that it could be a denial (if I send a `Approved: false`) or maybe something silly like `file-transfer-request-answer` ? 

I'll move these to `api/constants` since I didn't find a more reasonable spot.

**Review by capnspacehook — APPROVED** [review]
Looks good assuming all other feedback is addressed

File: api/constants/constants.go
Comment (avatus) [reply api/constants/constants.go:421]: I think I'll go with `Decision` since `Resolution` seems final (even tho more responses might be able to come in) and `Confirmation` to me seems too close to `Approve` but I'm probably over thinking. Anyway, `Decision` it is! Thanks for the brainstorm.

Comment (avatus) [reply api/constants/constants.go:421]: Before I go and change everything, are we saying change every instance of `FileTransferResponse` to `FileTransferDecision` or just for these SSH requests

File: lib/srv/regular/sshserver.go
Comment (avatus) [reply lib/srv/regular/sshserver.go:1648]: Does that need to happen sequentially? I'm not sure if that's the right word but what I mean is, other events or types of requests could be happening before someone would send a decision (or multiple decisions if the policy requires it).

Comment (avatus) [reply lib/srv/regular/sshserver.go:1648]: Ok cool, sounds good to me. I'll poke around and see how to refactor this.

File: api/proto/teleport/legacy/types/events/events.proto
Comment (avatus) [reply api/proto/teleport/legacy/types/events/events.proto:633]: Yes, this event was [created by me](https://github.com/gravitational/teleport/pull/24558) to support _this_ PR. so it's not consumed by anything yet.

**Review by rosstimothy — APPROVED** [review]

**Pull request merged**
